### PR TITLE
buildGoModule: Use the `env` attribute to pass environment variables 

### PR DESCRIPTION
**Pull request by ShamrockLee**
Pass all environment variables with `env` and instruct users to do so in the Nixpkgs Reference Manual.

This is part of the efforts to allow `buildGoModule` to take `__structuredAttrs = true`. This change would only rebuild the documentation ~and is easy to merge~ (it turns out to be a tree-wide fix, but fortunately, reviewers can reproduce the tree-wide changes by string substitution). Other significant/mass-rebuilding changes are split away to make reviewing smoother.

One may notice that the added code's indentation differs from the surrounding code. That is because the original code is out-of-formatting and indented too much.

These changes should rebuild no packages other than the documentation.

Cc:
@doronbehar @kalbasit @zowoq (people who might be interested in `buildGoModule`)
@emilazy @wolfgangwalther (people who might be interested in `__structuredAttrs`)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ShamrockLee** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 359641`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
    <li>promtail</li>
  </ul>
</details>

**Comment by emilazy** [comment]
Does it maybe make sense to keep `CGO_ENABLED` as part of the interface here like `GOFLAGS`, just because it’s so ubiquitous? (No strong opinion here, just wondering. Isn’t `GOFLAGS` also just an environment variable ultimately?)

**Comment by ShamrockLee** [comment]
> Does it maybe make sense to keep `CGO_ENABLED` as part of the interface here like `GOFLAGS`, just because it’s so ubiquitous?

It would be easier for now, as you are suggesting.

Still, if we plan to convert build helpers to support fixed-point attributes (`finalAttrs: { ... }`) like `stdenv.mkDerivation` does, the fewer `removeAttrs` the better. (Build helpers without `removeAttrs` can use `overrideAttrs` to achieve out-of-the-box conversion without waiting for #234651.)

**Comment by ShamrockLee** [comment]
> Edit: Ah, I see you have plans for `GOFLAGS` too, okay :)

#359744 is the plan.

File: pkgs/build-support/go/module.nix [hunk 1]
```diff
@@ -59,11 +59,13 @@
...
-  GO111MODULE = "on";
-  GOTOOLCHAIN = "local";
-
+  args =
+    removeAttrs args' (
+      [ "overrideModAttrs" "vendorSha256" ]
+      # Compatibility layer to the directly-specified CGO_ENABLED.
+      # TODO(@ShamrockLee): Remove after Nixpkgs 25.05 branch-off
+      ++ [ "CGO_ENABLED" ]
+    );
```
Comment (doronbehar) [inline pkgs/build-support/go/module.nix:65]: Nit:

```suggestion
  args = removeAttrs args' [
    "overrideModAttrs"
    "vendorSha256"
    # Compatibility layer to the directly-specified CGO_ENABLED.
    # TODO(@ShamrockLee): Remove after Nixpkgs 25.05 branch-off
    "CGO_ENABLED"
  ];
```

**Review by kalbasit — APPROVED** [review]
Thanks @ShamrockLee

**Review by doronbehar — APPROVED** [review]
Changes to `buildGoModule` look pretty good to me. Haven't iterated the packages' specific commits, but I guess they are trivial and should be good to go.

File: nixos/doc/manual/release-notes/rl-2505.section.md [hunk 2]
```diff
@@ -25,6 +25,12 @@
 - `binwalk` was updated to 3.1.0, which has been rewritten in rust. The python module is no longer available.
   See the release notes of [3.1.0](https://github.com/ReFirmLabs/binwalk/releases/tag/v3.1.0) for more information.
 
+- `buildGoModule` now pases environment variables via the `env` attribute. `CGO_ENABLED` should now be specified with `env.CGO_ENABLED` when passing to buildGoModule. Direct specification of `CGO_ENABLED` is now redirected by a compatibility layer with a warning, but will become an error in future releases.
```
Comment (zowoq) [inline nixos/doc/manual/release-notes/rl-2505.section.md:28]:
```suggestion
- `buildGoModule` now passes environment variables via the `env` attribute. `CGO_ENABLED` should now be specified with `env.CGO_ENABLED` when passing to buildGoModule. Direct specification of `CGO_ENABLED` is now redirected by a compatibility layer with a warning, but will become an error in future releases.
```

File: doc/languages-frameworks/go.section.md [hunk 3]
```diff
@@ -176,13 +176,13 @@ Following example could be used to only build the example-cli and example-server
 
 Specified as a string or list of strings. Causes the builder to skip building child packages that match any of the provided values.
 
-### `CGO_ENABLED` {#var-go-CGO_ENABLED}
+### `env.CGO_ENABLED` {#var-go-CGO_ENABLED}
```
Comment (katexochen) [inline doc/languages-frameworks/go.section.md:179]: There is a section about "Controlling the Go environment" (L248). Which is hinting at GOFLAGS and the other things you'd like to see in `env.`. Maybe we can somehow join these? Preferably move the that sec directly under the var section and env.CGO_ENABLED into it.

Comment (ShamrockLee) [reply doc/languages-frameworks/go.section.md:179]: Cool!

**Comment by ShamrockLee** [comment]
Regarding the use of environment variables to control the Go compiler, the current implementation of this PR shadows `env.GO111MODULES`, `env.GOTOOLCHAIN`, `env.GOOS`, and `env.GOARCH`, as the previous `buildGoModule` implementation did to the directly-specified version of these arguments.

Should we respect users' specifications for all these variables via `env` or keep shadowing them silently?

File: doc/languages-frameworks/go.section.md (hunk 3 above)
Comment (ShamrockLee) [reply doc/languages-frameworks/go.section.md]: I moved the paragraphs under `var-go-CGO_ENABLED` under the environment variables subsection and mentioned using the `env` attribute.

**Comment by ShamrockLee** [comment]
Rebased and resolved merge conflicts with #359798.

**Comment by TomaSajt** [comment]
Is there a plan to also move `GOFLAGS` in the future?
Currently, the nix logic adds additional values to `GOFLAGS`, but that could also be done from bash as well.
I did it like this: https://github.com/NixOS/nixpkgs/blob/9273a741451445a07834a4f2010f1040de0aaeef/pkgs/build-support/go/build-hook.sh#L6-L20

(This doesn't handle the current nix-side `GOFLAGS` validation `warnIf`s)

---
(I don't have a lot of experience with different variable types in bash, so I might have some misconceptions.)

**Comment by ShamrockLee** [comment]
> Is there a plan to also move `GOFLAGS` in the future?

It's #359744.

Once this PR is merged, I'll rebase #359744, dropping the `[placeholder] ...` commit, and put `GOFLAGS` into `env = { ... }`.

**Comment by ShamrockLee** [comment]
I rebased and resolved the merge conflict (seemingly caused by the recent treewide formatting).

This PR triggers no rebuild or cleanup except for the documentation-related packages, indicating that it works as intended. Do you think we should merge it now?

**Review by katexochen — APPROVED** [review]

**Comment by ShamrockLee** [comment]
ofborg-eval reports three new evaluation warnings after this change. (They are the warnings I added against the obsolete way of specifying the `CGO_ENABLED` environment variable.) I'll fix those packages before merging this PR.

This also informs us that the current GH-Action-based evaluation misses the additional warnings generated after the changes.

**Comment by ShamrockLee** [comment]
Rebased onto the top of `master` and resolved merge conflicts.

**Review by wolfgangwalther — APPROVED** [review]

**Comment by ShamrockLee** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 359641`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

**Comment by drupol** [comment]
Should we backport this?

**Comment by ShamrockLee** [comment]
> Should we backport this?

This PR itself is backward-incompatible.

If it is for the backporting of future PRs, some changes need to be re-implemented on the release branch.

**Pull request merged**